feat: Double tap to seek

retrozenith opened the pull request:
# 📦 Pull Request

## 🔖 Summary
Implemented the "Double Tap to Seek" (Corners Control) feature for the video player. This allows users to double-tap the left or right side of the screen to rewind or skip forward, respectively. The feature includes a "stacked seek" behavior where consecutive double-taps accumulate the seek time (e.g., +30s, +60s) and works seamlessly even when the player controls are visible.

## 🏷️ Ticket / Issue
- Feature: Corners Control (Double Tap to Seek)

## 🛠️ What’s Changed
- **feat(player)**: implement double tap to seek with stacked duration

## 📋 Details
This implementation adds standard mobile player gesture controls:
- **Left Double-Tap**: Rewind (uses configured `rewindSkipTime`, default 10s).
- **Right Double-Tap**: Skip Forward (uses configured `forwardSkipTime`, default 30s).
- **Stacked Seeking**: Repeatedly double-tapping accumulates the time difference in the visual feedback before committing the seek action.
- **Gesture Handling**: A 300ms delay was introduced to the single-tap action (toggling controls) to differentiate it from a double-tap. This is opt-in via the "Double tap to seek" setting.
- **Refactor**: The [GestureOverlay](cci:1://file:///home/cvictor/Code/streamyfin/components/video-player/controls/GestureOverlay.tsx:25:0-378:2) was refactored to remove an early return, ensuring that tap gestures are captured even when the video controls overlay is active/dimmed.

### ⚠️ Breaking Changes
None.

### 🔐 Security & Privacy Impact
None.

### ⚡ Performance Impact
- Animations use native drivers where possible.
- State updates for feedback animations are batched with `requestAnimationFrame` to maintain UI smoothness during gestures.

### 🖼️ Screenshots / GIFs (if UI)
- New toggle in **Settings -> Playback Controls**.
- Visual feedback overlay (`+/- 10s`) appears on the respective side of the player when activated.

![Screenshot_20251221_195331_Streamyfin](https://github.com/user-attachments/assets/39f6fbff-bd31-455b-988c-73afa62019f0)
![Screenshot_20251221_195440_Streamyfin](https://github.com/user-attachments/assets/0bc9ce93-8cef-442c-bc76-e9ac3b587795)

## ✅ Checklist
- [x] I’ve read the [contribution guidelines](CONTRIBUTING.md)
- [x] Code follows project style and passes lint/format (`bun run lint`)
- [x] Type checks pass (`bun run typecheck`)
- [ ] Docs updated (README/ADR/usage/API)
- [x] No secrets/credentials included; env vars documented
- [ ] Release notes/CHANGELOG entry added (if applicable)
- [x] Verified locally that changes behave as expected

## 🔍 Testing Instructions
1. Checkout the branch: `git checkout dt2s-feature`
2. Install dependencies: `bun install`
3. Run on a simulator/device: `bun run android` or `bun run ios`
4. **Verification Step 1 (Settings)**:
   - Go to Settings -> Playback Controls.
   - Enabling "Double tap to seek".
5. **Verification Step 2 (Playback)**:
   - Play any video.
   - Double-tap the right side: Should skip forward (e.g., +30s).
   - Double-tap the left side: Should rewind (e.g., -10s).
   - **Stacking**: Double-tap multiple times quickly on the same side. The feedback text should accumulate (e.g., +30s -> +60s).
   - **Controls Visibility**: Tap once to show controls, then double-tap. The seek action should still trigger correctly.
   - **Single Tap**: Single tapping should still toggle the controls (with a slight 300ms delay when the feature is enabled).

## ⚙️ Deployment Notes
- Standard React Native / Expo build process. No special native dependencies added.

## 📝 Additional Notes
- None

retrozenith committed:
feat: add corners control (dt2s)

retrozenith committed:
feat: implement stacked seek and add english base translation

retrozenith committed:
refactor(dt2s): apply PR suggestions from copilot review

fredrikburmester commented:
Is the slide to seek still there? Is there a setting to disable double tap to seek?

retrozenith commented:
Yeah, both are working at the same time on android at least, but using them in parallel might not be a good idea maybe making a toggle between the two ? or showing an warning ? I see no issue but might cause confusion to end-user

fredrikburmester commented:
Yes we should add a dropdown option selecting between the two.

retrozenith commented:
on it. Also didn't expect copilot to make code review so well.

retrozenith committed:
feat(settings): add Smart Gestures dropdown menu

- Create dropdown menu grouping double-tap and slide to seek features
- Add Smart Gestures translation keys to en.json
- Replace individual toggle switches with PlatformDropdown component
- Display feature descriptions within dropdown options
- Show current state in trigger (Both enabled, 1 enabled, Disabled)
- Improve UX by consolidating related gesture settings

retrozenith committed:
feat(settings): add conflict warning for Smart Gestures

- Display warning when both double-tap and slide to seek are enabled
- Add smart_gestures_warning translation key
- Show yellow warning banner below dropdown when both features active
- Inform users features weren't designed for simultaneous use

retrozenith committed:
refactor(settings): replace dropdown with section headers for Smart Gestures

- Remove PlatformDropdown in favor of section headers with individual switches
- Group gestures into Smart Gestures and Other Gestures sections
- Add Alert popup warning when both smart gestures are enabled
- Ensure UI properly updates when toggling switches
- Improve visual organization with section headers

retrozenith commented:
https://github.com/user-attachments/assets/f7e1a478-087c-4409-8930-91c416171929

there's an issue if we use the dropdown way.

retrozenith committed:
refactor(settings): replace Alert popup with yellow warning banner

- Remove Alert popup for Smart Gestures conflict
- Restore yellow warning banner below toggles
- Update warning message to be more concise
- Show warning when both features are enabled

retrozenith commented:
yellow warning looks way better when compared to popup as well. Here's the settings preview of changes

![Screenshot_20251223_200631_Streamyfin](https://github.com/user-attachments/assets/b4cd6446-d827-46db-adbf-4d929e812268)
![Screenshot_20251223_200640_Streamyfin](https://github.com/user-attachments/assets/d55a360f-8951-4658-8922-280af838e780)

fredrikburmester commented:
How did you get around showing the controls on single press? They are not conflicting? That was my main issue.

fredrikburmester commented:
Please do a dropdown instead. We don't want users to have both enabled at the same time, that seems unnecessary. No?

retrozenith commented:
1. I delayed tap detection so that when you first press it waits 300ms if second tap comes then surely it's dt2s if not displays controls.

from useGestureDetection.ts
`const DOUBLE_TAP_DELAY_MS = 300;`

2. let me try

retrozenith committed:
feat(settings): add Seek Gestures dropdown translation keys

- Add gesture_mode_none, gesture_mode_double_tap, gesture_mode_swipe keys
- Update smart_gestures label to 'Seek Gestures'
- Update description to 'Choose your preferred seek method'

retrozenith committed:
feat(ui): enhance PlatformDropdown to support multi-line labels with styled descriptions

- Split labels by newline character to detect title and description
- Render first line as white title text
- Render subsequent lines as gray description text (text-neutral-500)
- Apply smaller font size to description text
- Maintain backward compatibility with single-line labels

retrozenith committed:
feat(settings): implement single-select Seek Gestures dropdown

- Replace separate toggles with single-select dropdown for seek gestures
- Add radio options: None, Double Tap to Seek, Horizontal Swipe Skip
- Ensure only one seek gesture can be active at a time
- Remove section headers for cleaner UI
- Remove warning banner (no longer needed with mutually exclusive selection)
- Add descriptions to each dropdown option for better UX
- Keep Brightness and Volume controls as normal toggles
- Maintain plugin settings lock compatibility

retrozenith commented:
idk if it should be kept or no maybe making a poll would be a good idea. In any case if you want to get back to how it was with both at the same time then just reset HEAD to commit [d1686341679bfd8cbc66ce8052fede70d19f4333](https://github.com/streamyfin/streamyfin/pull/1289/commits/d1686341679bfd8cbc66ce8052fede70d19f4333) and push

retrozenith committed:
refactor: Remove detailed gesture control descriptions from translations and UI labels as they don't display properly on iOS

retrozenith committed:
refactor: shorten gesture setting labels.

fredrikburmester commented:
I tried the 300ms delay myself a while ago and i though that made the controls feel slow.. But maybe that's just me..

retrozenith commented:
After testing for a week. I agree with you maybe 230-250 ?

retrozenith committed:
Merge branch 'develop' into dt2s-feature

retrozenith committed:
tune(ui): adjust double tap detection delay for responsiveness to 230ms

retrozenith commented:
After testing i guess it is 230ms, but i think there is some more work needed for dt2s stacking to make it a bit more smooth like as long as overlay is displayed any tap on the right side will be added to stack for forward seeking

retrozenith committed:
Merge branch 'develop' into dt2s-feature

retrozenith committed:
Merge branch 'develop' into dt2s-feature

retrozenith committed:
Merge branch 'upstream/develop' into dt2s-feature
